### PR TITLE
Add in transcript quantification processes and BAM input option

### PR DESCRIPTION
**Pull request by yuukiiwa**
## PR checklist

- [x] This comment contains a description of changes (with reason)
   - Reason: The nanoseq pipeline did not have a BAM input option and a transcript quantification process for `cDNA` and `directRNA` protocols.
   - 1st change: Add in a BAM input option
      - Now the pipeline accepts both fastq and bam files as inputs from the samplesheet.csv `input_file` column (previously named `fastq`)
   - 2nd change: Add in a transcript quantification process
      - transcript quantification method (either [`bambu`](https://github.com/GoekeLab/bambu)(default) or [`stringtie2`](https://ccb.jhu.edu/software/stringtie/)) can be specified by the `--transcriptquant` flag 
        - the `bambu` option uses [`bambu`](https://github.com/GoekeLab/bambu) for both transcript reconstruction and quantification
        - the `stringtie2` option uses [`stringtie2`](https://ccb.jhu.edu/software/stringtie/) for transcript reconstruction and [`featureCounts`](http://subread.sourceforge.net/) for quantification
        - both options output a final `*.gtf` file with extended annotation, and `count_gene.txt` and `count_transcript.txt` files with gene and transcript counts for each sample
      - Differential expression analyses on genes and transcripts are done by DESeq2 and DEXSeq, respectively
         - requires at least two conditions stated in the samplesheet.csv`condition` column and at least three samples in each of the conditions
      - Added in containers in base.config for `StringTie2`, `FeatureCounts`, and `Bambu` and changed the `process.container` to `yuukiiwa/nanoseq:0.1` with all dependencies.
- [x] If you've fixed a bug or added code that should be tested, add tests!
   - [ ] If necessary, also make a PR on the [nf-core/nanoseq branch on the nf-core/test-datasets repo](https://github.com/nf-core/test-datasets/pull/new/nf-core/nanoseq)
      - Test datasets used are from the [NanoporeRNASeq](https://github.com/GoekeLab/NanoporeRNASeq) package
      - Test datasets used and corresponding samplesheets can be downloaded [here](https://drive.google.com/file/d/1knfC-ALiv8RPJhoztnQc3MLg_6-yVhGR/view?usp=sharing)
   - [ ] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [x] Documentation in `docs` is updated
   - `usage.md`
       - Added `--transcriptquant` and `--skip_transcriptquant` flags for transcript quantification processes
       - Added an additional `condition` column in the samplesheet.csv for differential expression analysis purpose.
       - Changed the name of the`fastq` column to `input_file` to accommodate for BAM inputs.
       - Added in a sample command for running with BAM inputs.
   - `output.md`
      - Added in `<ALIGNER>/bam` and `<ALIGNER>/bam_index` directories in **Alignment**
      - Added in a **Transcript Reconstruction and Quantification** section
- [x] `README.md` is updated
   - Added descriptions on transcript quantification processes in the pipeline summary.
   - Made changes in contributors.
- [ ] Make sure your code lints (`nf-core lint .`).
   - missing `nextflow_schema.json` causes `check_files_existERROR`
- [ ] `CHANGELOG.md` is updated

### CODE REVIEW

File: bin/check_samplesheet.py [hunk 1]
```diff
@@ -20,7 +20,7 @@ def print_error(error,line):
 
 
 def check_samplesheet(FileIn,FileOut):
-    HEADER = ['sample', 'fastq', 'barcode', 'genome', 'transcriptome']
+    HEADER = ['sample', 'sample_path', 'barcode', 'genome', 'transcriptome','condition']
```
Comment (drpatelh) [inline bin/check_samplesheet.py:23]:
```suggestion
    HEADER = ['sample', 'input_file', 'barcode', 'genome', 'transcriptome','condition']
```

Comment (drpatelh) [reply bin/check_samplesheet.py:23]: `input_file` is a bit more generic and sounds better 👍

File: conf/base.config [hunk 2]
```diff
@@ -60,6 +60,18 @@ process {
...
+  }
+  withName:StringTie2Merge {
+    container = 'quay.io/biocontainers/stringtie:2.0--hc900ff6_0'
+  }
+  withName:FeatureCounts {
+    container = 'quay.io/biocontainers/subread:2.0.1--hed695b0_0'
+  }
+  withName:Bambu {
+    container = 'docker.io/cheny1/bambu:0.3'
+  }
```
Comment (drpatelh) [inline conf/base.config:74]: Add all of this software to `environment.yml`

File: docs/usage.md [hunk 3]
```diff
@@ -149,10 +152,11 @@ You will need to create a file with information about the samples in your experi
 | Column          | Description                                                                                                                |
 |-----------------|----------------------------------------------------------------------------------------------------------------------------|
 | `sample`        | Sample name without spaces.                                                                                                |
-| `fastq`         | Full path to FastQ file if previously demultiplexed. File has to be zipped and have the extension ".fastq.gz" or ".fq.gz". |
+| `sample_path`   | Full path to FastQ file if previously demultiplexed or to BAM file if previously aligned. FastQ File has to be zipped and have the extension ".fastq.gz" or ".fq.gz". BAM file has to have the extension ".bam". |
```
Comment (drpatelh) [inline docs/usage.md:155]:
```suggestion
| `input_file`   | Full path to FastQ file if previously demultiplexed or to BAM file if previously aligned. FastQ File has to be zipped and have the extension ".fastq.gz" or ".fq.gz". BAM file has to have the extension ".bam". |
```

Comment (drpatelh) [reply docs/usage.md:155]: Needs to be changed everywhere.

File: environment.yml [hunk 4]
```diff
@@ -24,3 +24,11 @@ dependencies:
...
   - bioconda::ucsc-bedgraphtobigwig=357
   - bioconda::ucsc-bedtobigbed=357
+  
+  - r-base=4.0.1
+  - r-devtools=2.3.0
+  - bioconductor-deseq2=1.28.0
+  - bioconductor-drimseq=1.16.0
+  - bioconductor-dexseq=1.34.0
+  - bioconductor-stager=1.10.0
+  - bioconductor-bsgenome=1.56.0
```
Comment (drpatelh) [inline environment.yml:34]:
```suggestion
  - conda-forge::r-base=4.0.1
  - conda-forge::r-devtools=2.3.0
  - bioconda::bioconductor-deseq2=1.28.0
  - bioconda::bioconductor-drimseq=1.16.0
  - bioconda::bioconductor-dexseq=1.34.0
  - bioconda::bioconductor-stager=1.10.0
  - bioconda::bioconductor-bsgenome=1.56.0
```

Comment (drpatelh) [reply environment.yml:34]: I have created a separate section for `conda-forge` packages at the top of the file.

File: nextflow.config [hunk 5]
```diff
@@ -79,7 +83,7 @@ params {
 
 // Container slug. Stable releases should specify release tag!
 // Developmental code should specify :dev
-process.container = 'nfcore/nanoseq:dev'
+process.container = 'yuukiiwa/nanoseq:0.1'
```
Comment (drpatelh) [inline nextflow.config:86]:
```suggestion
process.container = 'nfcore/nanoseq:dev'
```

Comment (drpatelh) [reply nextflow.config:86]: This has to stay as `nfcore/nanoseq:dev` so we will have to update `environment.yml` with the required packages in order to build this container.

File: conf/base.config (hunk 2 above)
Comment (cying111) [reply conf/base.config:74]: for others, we can add it to environment.yml, but for bambu, there is no existing software published any where yet(in fact we plan to submit to bioconductor later), so what should we do with that?

Comment (drpatelh) [reply conf/base.config:74]: Ok. You can leave that in there then 👍 The container just needs to be tagged properly by release for reproducibility.

File: bin/check_samplesheet.py (hunk 1 above)
Comment (yuukiiwa) [reply bin/check_samplesheet.py]: sample_path has been changed to input_file.

File: docs/usage.md (hunk 3 above)
Comment (yuukiiwa) [reply docs/usage.md]: sample_path has been changed to input_file.

File: environment.yml (hunk 4 above)
Comment (yuukiiwa) [reply environment.yml]: conda-forge and bioconda packages are now separated.

File: nextflow.config (hunk 5 above)
Comment (yuukiiwa) [reply nextflow.config]: container has been changed to nfcore/nanoseq:dev

**Comment by yuukiiwa** [comment]
Hi @drpatelh, we have tried solving the merge conflicts by reverting README.md and nextflow_schema.json to their nf-core/nanoseq-dev versions. The two failing tests are from linting of nextflow_schema.json, which currently lacks several parameters (Currently, I can't do any changes on my side as it creates merge conflicts), and CHANGELOG.md updated with the dependencies from the docker-image update PR.

**Comment by drpatelh** [comment]
Thanks @yuukiiwa ! I will have to look at this PR in a bit more detail because it appears that some of the changes I have introduced via the `TEMPLATE` merge when nf-core/tools was released have been reverted back to the older versions. I probably won't have time to do that right now so I will merge this now and come back to fix the changes all at once. In the meantime, it would be good if you can play with the `dev` version of the code and create additional PRs for any other functionality you need to.

**Review by drpatelh — APPROVED** [review]
Thanks @yuukiiwa @cying111 😎

**Comment by yuukiiwa** [comment]
Thanks, @drpatelh!

**Pull request merged**